net/ip.c

```diff
@@ -0,0 +1,448 @@
+				break;
+		}
+	}
+	if (f >= ARRAY_SIZE(ip_family) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	for (s = 0; s < ARRAY_SIZE(ip_protocol); ++s) {
+		if (ip_protocol[s] != NULL) {
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(ep_name, ip_protocol[s], shift) == 0)
+				break;
+		}
+	}
+	if (s >= ARRAY_SIZE(ip_protocol) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	at = strchr(ep_name, '@');
+	if (at == NULL)
+		return M0_ERR(-EINVAL);
+	else {
+		at++;
+		if (at == NULL || at[0] < '0' || at[0] > '9')
+			return M0_ERR(-EINVAL);
+		strcpy(port, at);
+		portnum = atoi(port);
+		M0_ASSERT(portnum < M0_NET_IP_PORT_MAX);
+		addr->nia_n.nip_port = (uint16_t)portnum;
+	}
+
+	rc = m0_net_hostname_to_ip((char *)ep_name, ip, &addr->nia_n.nip_format);
+	if (rc == 0)
+		inet_pton(f == M0_NET_IP_AF_INET ? AF_INET : AF_INET6,
+			  ip, &addr->nia_n.ip_n.sn[0]);
+
+	M0_ASSERT(strlen(name) < ARRAY_SIZE(addr->nia_p));
+	strcpy(addr->nia_p, name);
+	addr->nia_n.fmt_pvt.ia.nia_family = f;
+	addr->nia_n.fmt_pvt.ia.nia_type = s;
+
+	/* Ignore the error due to gethostbyname() as it will be retried. */
+	return rc >= 0 ? M0_RC(0) : M0_ERR(rc);
+}
+
+/**
+ * This function decodes the lnet format address and extracts the ip address and
+ * port number from it.
+ * This is also used to allocate unique transfer machine identifiers for LNet
+ * network addresses with wildcard transfer machine identifier (like
+ * "192.168.96.128@tcp:12345:31:*").
+ * Return value: 0 in case of success.
+ *               < 0 in case of error.
+ */
+static int m0_net_ip_lnet_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	char        *at = NULL;
+	int          nr;
+	int          i;
+	unsigned     pid;
+	unsigned     portal;
+	unsigned     portnum;
+	unsigned     tmid;
+	char         node[M0_NET_IP_STRLEN_MAX] = {};
+	char         port[M0_NET_IP_PORTLEN_MAX] = {};
+	const char  *ep_name = name;
+	uint32_t     nia_n;
+	int          shift;
+	int          s;
+
+	at = strchr(ep_name, '@');
+	if (strncmp(ep_name, "0@lo", 4) == 0) {
+		nia_n = htonl(INADDR_LOOPBACK);
+		inet_ntop(AF_INET, &nia_n, node, ARRAY_SIZE(node));
+	} else {
+		if (at == NULL || at - ep_name >= sizeof node)
+			return M0_ERR(-EPROTO);
+
+		M0_PRE(sizeof node >= (at-ep_name)+1);
+		memcpy(node, ep_name, at - ep_name);
+	}
+	at++;
+
+	for (s = 0; s < ARRAY_SIZE(ip_protocol); s++) {
+		if (ip_protocol[s] != NULL) {
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(at, ip_protocol[s], shift) == 0)
+			{
+				addr->nia_n.fmt_pvt.la.nla_type = s;
+				break;
+			}
+		}
+	}
+
+	if (s == ARRAY_SIZE(ip_protocol) || ((at = strchr(at, ':')) == NULL))
+		return M0_ERR(-EPROTO);
+
+	nr = sscanf(at + 1, "%u:%u:%u", &pid, &portal, &tmid);
+	if (nr != 3) {
+		nr = sscanf(at + 1, "%u:%u:*", &pid, &portal);
+		if (nr != 2)
+			return M0_ERR(-EPROTO);
+		for (i = 0; i < ARRAY_SIZE(ip_autotm); ++i) {
+			if (ip_autotm[i] == 0) {
+				tmid = i;
```

Comment:
If there are two or more threads are trying to parse an address then we have a race condition here. Probably, we should protect the global variable with a lock.

Reply:
Fixed by adding mutex lock where `ip_autotm[]` is getting manipulated.